src/vcpkg/base/files.cpp

```diff
@@ -1453,25 +1455,64 @@ namespace vcpkg
         std::string output;
         constexpr std::size_t buffer_size = 1024 * 32;
         char buffer[buffer_size];
-        do
+        auto this_read = this->read(buffer, 1, buffer_size);
+        if (this_read != buffer_size)
         {
-            const auto this_read = this->read(buffer, 1, buffer_size);
-            if (this_read != 0)
+            auto maybe_error = ::ferror(m_fs);
+            if (maybe_error)
             {
-                output.append(buffer, this_read);
+                ec.assign(maybe_error, std::generic_category());
+                return output;
             }
-            else if ((ec = this->error()))
+
+            if (!this->eof())
             {
-                return std::string();
+                Checks::unreachable(VCPKG_LINE_INFO, "Got a partial read without an error or end");
             }
-        } while (!this->eof());
+        }
 
-        if (Strings::starts_with(output, "\xEF\xBB\xBF"))
         {
-            // remove byte-order mark from the beginning of the string
-            output.erase(output.begin(), output.begin() + 3);
+            const char* to_append = buffer;
+            size_t to_append_size = this_read;
+            if (to_append_size >= 3 && ::memcmp(to_append, "\xEF\xBB\xBF", 3) == 0)
+            {
+                // remove byte-order mark from the beginning of the string
+                to_append_size -= 3;
+                to_append += 3;
+            }
+
+            output.append(to_append, to_append_size);
+        }
+
+        if (this_read == buffer_size)
+        {
+            for (;;)
+            {
+                this_read = this->read(buffer, 1, buffer_size);
+                if (this_read != buffer_size)
+                {
+                    break;
+                }
+
+                output.append(buffer, this_read);
+            }
+
+            auto maybe_error = ::ferror(m_fs);
+            if (maybe_error)
+            {
+                ec.assign(maybe_error, std::generic_category());
+                output.clear();
+                return output;
+            }
+
+            output.append(buffer, this_read);
+            if (!this->eof())
+            {
+                Checks::unreachable(VCPKG_LINE_INFO, "Got a partial read without an error or end");
+            }
         }
```

Comment:
Could this common chunk of "append the rest of the stream without transformation" be deduplicated between `read_to_end` and `best_effort_read_contents_if_shebang`?